### PR TITLE
chore: create user profile

### PR DESCRIPTION
**Pull request by Sboonny**
- [x] I have read [Chapter's contributing guidelines](https://github.com/freeCodeCamp/chapter/blob/main/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update README.md`).
- [x] My pull request targets the `main` branch of Chapter.

In this attempt, I won't try to add the ability to get more data, but will focus on deletion function.

### CODE REVIEW

File: client/src/modules/profiles/pages/userProfile.tsx [hunk 1]
```diff
@@ -0,0 +1,66 @@
+import React from 'react';
+import { Flex, Heading, Text, Link } from '@chakra-ui/layout';
+import { useConfirmDelete } from 'chakra-confirm';
+import { useRouter } from 'next/router';
+import { Button } from '@chakra-ui/button';
+import { useDeleteMeMutation } from 'generated/graphql';
+import { useAuth } from 'modules/auth/store';
```
Comment (ojeytonwilliams) [inline client/src/modules/profiles/pages/userProfile.tsx:7]:
```suggestion
import { useAuthStore } from 'modules/auth/store';
```
By using the store we can update the stored user after deleteMe resolves.

File: client/src/modules/profiles/pages/userProfile.tsx [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+import { Users } from 'modules/dashboard/Users/graphql/queries';
+
+export const UserProfilePage = () => {
+  const { user } = useAuth();
```
Comment (ojeytonwilliams) [inline client/src/modules/profiles/pages/userProfile.tsx:11]:
```suggestion
  const {
    data: { user },
    setData,
  } = useAuthStore();
```

File: client/src/modules/profiles/pages/userProfile.tsx [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+  const confirmDelete = useConfirmDelete({ doubleConfirm: true });
+  const [deleteMe] = useDeleteMeMutation({
+    refetchQueries: [{ query: Users }],
+  });
+  const clickDelete = async () => {
+    const ok = await confirmDelete();
+    if (!ok) return;
+    deleteMe({ variables: { userId } });
+    router.push('/');
```
Comment (ojeytonwilliams) [inline client/src/modules/profiles/pages/userProfile.tsx:27]:
```suggestion
    setData({});
    router.push('/');
```
Purge the stored user.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]

File: client/src/modules/auth/graphql/mutations.ts [hunk 4]
```diff
@@ -0,0 +1,9 @@
+import { gql } from '@apollo/client';
+
+export const deleteMe = gql`
+  mutation deleteMe($userId: Int!) {
+    deleteMe(id: $userId) {
```
Comment (ojeytonwilliams) [inline client/src/modules/auth/graphql/mutations.ts:5]:
```suggestion
  mutation deleteMe {
    deleteMe {
```
By giving this an id parameter it implies that you can delete a user with that id.  That's not the case, so we should remove the parameter.

File: server/src/controllers/Auth/resolver.ts [hunk 5]
```diff
@@ -1,12 +1,21 @@
...
 @Resolver()
 export class AuthResolver {
   @Query(() => UserWithInstanceRole, { nullable: true })
   async me(@Ctx() ctx: ResolverCtx): Promise<UserWithInstanceRole | null> {
     return ctx.user ?? null;
   }
+
+  @Mutation(() => User)
+  async deleteMe(
+    @Arg('id', () => Int) id: number,
```
Comment (ojeytonwilliams) [inline server/src/controllers/Auth/resolver.ts:16]:
```suggestion
```
The queries should not be passing this id.

File: server/src/controllers/Auth/resolver.ts [hunk 6]
```diff
@@ -1,12 +1,21 @@
...
+    @Ctx() ctx: Required<ResolverCtx>,
+  ): Promise<User | undefined> {
+    if (id === ctx.user.id) return await prisma.users.delete({ where: { id } });
```
Comment (ojeytonwilliams) [inline server/src/controllers/Auth/resolver.ts:19]:
```suggestion
    return await prisma.users.delete({ where: { id: ctx.user.id } });
```

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
I noticed, by chance, that banned users can't delete their accounts.  For the MVP I think we can allow that (even knowing that they can create new accounts to evade bans) and consider a proper solution later down the line. Could you make the `user_bans` cascade like the sessions?

**Review by gikf — COMMENTED** [review]
Self note for future: remember to apply schema changes before wondering why updated relations do not cascade on deletion :eyes:

**Review by ojeytonwilliams — APPROVED** [review]
LGTM :+1:

**Pull request merged**